kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PaginatedScanSuite.scala

```diff
@@ -0,0 +1,437 @@
+      testName: String): Unit = {
+    val lastReadLogFilePath = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReadLogFilePath
+    val lastReturnedRowIndex = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReturnedRowIndex
+
+    assert(lastReadLogFilePath.endsWith(expectedLogFilePath))
+    assert(lastReturnedRowIndex == expectedRowIndex)
+
+    logger.info(s"$testName: New PageToken: lastReadLogFileName = $lastReadLogFilePath")
+    logger.info(s"$testName: New PageToken: lastReadRowIndex = $lastReturnedRowIndex")
+  }
+
+  private def runSinglePaginationTestCase(
+      testName: String,
+      tablePath: String,
+      tableVersionOpt: Optional[Long],
+      pageSize: Long,
+      expectedFileCount: Int,
+      expectedBatchCount: Int,
+      expectedLogFileName: String,
+      expectedRowIndex: Int): Unit = {
+    val paginatedScan = createPaginatedScan(
+      tablePath = tablePath,
+      tableVersionOpt = tableVersionOpt,
+      pageSize = pageSize)
+
+    val paginatedIter = paginatedScan.getScanFiles(customEngine)
+    val returnedBatchesInPage = collectPaginatedBatches(paginatedIter)
+
+    validatePageResults(returnedBatchesInPage, expectedFileCount, expectedBatchCount, testName)
+    validatePageToken(paginatedIter, expectedLogFileName, expectedRowIndex, testName)
+
+    paginatedIter.close()
+  }
+
+  // TODO: test call hasNext() twice
+
+  // File name constants for better readability
+  private val JSON_FILE_0 = "00000000000000000000.json"
+  private val JSON_FILE_1 = "00000000000000000001.json"
+  private val JSON_FILE_2 = "00000000000000000002.json"
+  private val JSON_FILE_11 = "00000000000000000011.json"
+  private val JSON_FILE_12 = "00000000000000000012.json"
+  private val CHECKPOINT_FILE_10 = "00000000000000000010.checkpoint.parquet"
+
+  // Single JSON file test cases
+  val singleJsonTestCases = Seq(
+    SinglePageRequestTestCase(
+      pageSize = 1,
+      expFileCnt = 2,
+      expBatchCnt = 1,
+      expLogFile = JSON_FILE_0,
+      expRowIdx = 4),
+    SinglePageRequestTestCase(
+      pageSize = 2,
+      expFileCnt = 2,
+      expBatchCnt = 1,
+      expLogFile = JSON_FILE_0,
+      expRowIdx = 4),
+    SinglePageRequestTestCase(
+      pageSize = 4,
+      expFileCnt = 5,
+      expBatchCnt = 2,
+      expLogFile = JSON_FILE_0,
+      expRowIdx = 7),
+    SinglePageRequestTestCase(
+      pageSize = 7,
+      expFileCnt = 5,
+      expBatchCnt = 2,
+      expLogFile = JSON_FILE_0,
+      expRowIdx = 7),
+    SinglePageRequestTestCase(
+      pageSize = 20,
+      expFileCnt = 5,
+      expBatchCnt = 2,
+      expLogFile = JSON_FILE_0,
+      expRowIdx = 7))
+
+  singleJsonTestCases.foreach { testCase =>
+    test(s"Single JSON file - page size ${testCase.pageSize}") {
+      withKernelStaticTable("kernel-pagination-all-jsons") { tablePath =>
+        /**
+         *  Log Segment List:
+         *  00000000000000000000.json contains 2 batches, 5 active AddFiles in total
```

Comment:
probably worth mentioning that this assumes, and requires, that the batch size is 5